File: cc/057.cc
// Project Euler Problem 057
//
// Investigate the expansion of the continued fraction for the square
// root of two.

#include <cstdio>

#include "bigmath.h"

/*
1 + 1/2 = 3/2
1 + 1/(2 + 1/2) = 7/5
1 + 1/(2 + 1/(2 + 1/2)) = 17/12
1 + 1/(2 + 1/(2 + 1/(2 + 1/2))) = 41/29

Not counting the leading ones, we have a simple pattern:

1/2
2/5
5/12
12/29
29/70
70/169
169/408
*/

int main(int argc, const char* argv[]) {
  int count = 0;
  int hits = 0;
  Bignum10 top(0);
  Bignum10 bottom(1);
  while (++count <= 1000) {
    Bignum10 temp = bottom;
    bottom *= 2;
    bottom += top;
    top = temp;
    temp += bottom;
    if (temp.size() > bottom.size()) {
      hits++;
    }
  }
  printf("%d\n", hits);
}

File: go/problems/057.go
// Project Euler Problem 057
//
// Investigate the expansion of the continued fraction for the square
// root of two.

package problems

import (
	"fmt"
	"math/big"
)

func Problem057() string {
	count := 0
	numerator, denominator, tmp := big.NewInt(3), big.NewInt(2), big.NewInt(0)
	for i := 0; i < 1000; i++ {
		tmp.Set(numerator)
		numerator.Add(numerator, denominator)
		numerator.Add(numerator, denominator)
		denominator.Add(denominator, tmp)
		if len(numerator.String()) > len(denominator.String()) {
			count++
		}
	}
	return fmt.Sprint(count)
}

func init() {
	Functions[57] = Problem057
}
